File: js-src/pieceFactory.js
import { Piece } from './piece.js';
import { PieceView } from './pieceView.js';

/**
 * Factory class for creating different types of Tetris pieces.
 * Converted from Java PieceFactory class.
 */
export class PieceFactory {
  /**
   * Creates an I-piece (straight line).
   * @returns {Piece} The I-piece
   */
  static createIPiece() {
    const shape = [
      [true, true, true, true]
    ];
    return new Piece(shape);
  }

  /**
   * Creates an O-piece (square).
   * @returns {Piece} The O-piece
   */
  static createOPiece() {
    const shape = [
      [true, true],
      [true, true]
    ];
    return new Piece(shape);
  }

  /**
   * Creates a T-piece.
   * @returns {Piece} The T-piece
   */
  static createTPiece() {
    const shape = [
      [false, true, false],
      [true, true, true]
    ];
    return new Piece(shape);
  }

  /**
   * Creates an L-piece.
   * @returns {Piece} The L-piece
   */
  static createLPiece() {
    const shape = [
      [true, false],
      [true, false],
      [true, true]
    ];
    return new Piece(shape);
  }

  /**
   * Creates a view for an I-piece.
   * @returns {PieceView} The I-piece view
   */
  static createIPieceView() {
    return new PieceView('I');
  }

  /**
   * Creates a view for an O-piece.
   * @returns {PieceView} The O-piece view
   */
  static createOPieceView() {
    return new PieceView('O');
  }

  /**
   * Creates a view for a T-piece.
   * @returns {PieceView} The T-piece view
   */
  static createTPieceView() {
    return new PieceView('T');
  }

  /**
   * Creates a view for an L-piece.
   * @returns {PieceView} The L-piece view
   */
  static createLPieceView() {
    return new PieceView('L');
  }
}

File: python_src/piece_factory.py
"""PieceFactory class - creates different types of Tetris pieces."""

from .piece import Piece
from .piece_view import PieceView


class PieceFactory:
    """Factory class for creating Tetris pieces and their views."""
    
    @staticmethod
    def create_i_piece() -> Piece:
        """Create an I-shaped piece (straight line).
        
        Returns:
            I-piece instance
        """
        shape = [
            [True, True, True, True]
        ]
        return Piece(shape)
    
    @staticmethod
    def create_o_piece() -> Piece:
        """Create an O-shaped piece (square).
        
        Returns:
            O-piece instance
        """
        shape = [
            [True, True],
            [True, True]
        ]
        return Piece(shape)
    
    @staticmethod
    def create_t_piece() -> Piece:
        """Create a T-shaped piece.
        
        Returns:
            T-piece instance
        """
        shape = [
            [False, True, False],
            [True, True, True]
        ]
        return Piece(shape)
    
    @staticmethod
    def create_l_piece() -> Piece:
        """Create an L-shaped piece.
        
        Returns:
            L-piece instance
        """
        shape = [
            [True, False],
            [True, False],
            [True, True]
        ]
        return Piece(shape)
    
    @staticmethod
    def create_i_piece_view() -> PieceView:
        """Create a view for I-piece.
        
        Returns:
            PieceView for I-piece
        """
        return PieceView('I')
    
    @staticmethod
    def create_o_piece_view() -> PieceView:
        """Create a view for O-piece.
        
        Returns:
            PieceView for O-piece
        """
        return PieceView('O')
    
    @staticmethod
    def create_t_piece_view() -> PieceView:
        """Create a view for T-piece.
        
        Returns:
            PieceView for T-piece
        """
        return PieceView('T')
    
    @staticmethod
    def create_l_piece_view() -> PieceView:
        """Create a view for L-piece.
        
        Returns:
            PieceView for L-piece
        """
        return PieceView('L')